Source/WebCore/platform/graphics/skia/SkiaSystemFallbackFontCache.cpp

```diff
@@ -0,0 +1,319 @@
+        slant = FC_SLANT_ITALIC;
+        break;
+    case SkFontStyle::kOblique_Slant:
+        slant = FC_SLANT_OBLIQUE;
+        break;
+    }
+}
+
+static String filePathFromPattern(FcPattern* pattern)
+{
+    FcChar8* filename = nullptr;
+    if (FcPatternGetString(pattern, FC_FILE, 0, &filename) != FcResultMatch || !filename)
+        return { };
+
+    const char* sysroot = reinterpret_cast<const char*>(FcConfigGetSysRoot(nullptr));
+    if (!sysroot)
+        return String::fromUTF8(reinterpret_cast<const char*>(filename));
+
+    return FileSystem::pathByAppendingComponent(String::fromUTF8(sysroot), String::fromUTF8(reinterpret_cast<const char*>(filename)));
+}
+
+struct FontSetCacheKey {
+    FontSetCacheKey() = default;
+
+    FontSetCacheKey(int weight, int width, int slant)
+        : weight(weight)
+        , width(width)
+        , slant(slant)
+    {
+    }
+
+    explicit FontSetCacheKey(WTF::HashTableDeletedValueType)
+        : weight(-1)
+        , width(-1)
+        , slant(-1)
+    {
+    }
+
+    bool isHashTableDeletedValue() const { return weight == -1 && width == -1 && slant == -1; }
+
+    bool operator==(const FontSetCacheKey&) const = default;
+
+    int weight { 0 };
+    int width { 0 };
+    int slant { 0 };
+};
+
+inline void add(Hasher& hasher, const FontSetCacheKey& key)
+{
+    add(hasher, key.weight, key.width, key.slant);
+}
+
+struct FontSetCacheKeyHash {
+    static unsigned hash(const FontSetCacheKey& key) { return computeHash(key); }
+    static bool equal(const FontSetCacheKey& a, const FontSetCacheKey& b) { return a == b; }
+    static const bool safeToCompareToEmptyOrDeleted = true;
+};
+
+class FontSetCache {
+    WTF_DEPRECATED_MAKE_FAST_ALLOCATED(FontSetCache);
+    WTF_MAKE_NONCOPYABLE(FontSetCache);
+public:
+    FontSetCache() = default;
+    ~FontSetCache() = default;
+
+    struct Font {
+        Font(String&& path, int ttcIndex, FcCharSet* charSet)
+            : path(WTFMove(path))
+            , ttcIndex(ttcIndex)
+            , charSet(charSet)
+        {
+        }
+
+        String path;
+        int ttcIndex { 0 };
+        FcCharSet* charSet { nullptr }; // Owned by FcFontSet.
+    };
+
+    const FontSetCache::Font* fontForCharacterCluster(const SkFontStyle& style, const String& locale, FcCharSet* charSet)
+    {
+        int weight, width, slant;
+        fontconfigStyle(style, weight, width, slant);
+
+        auto& fontSet = m_cache.ensure(FontSetCacheKey(weight, width, slant), [&] {
+            return FontSet::create(locale, weight, width, slant);
+        }).iterator->value;
+        if (!fontSet)
+            return nullptr;
+
+        return fontSet->bestForCharacterCluster(charSet);
+    }
+
+private:
+    class FontSet {
+        WTF_DEPRECATED_MAKE_FAST_ALLOCATED(FontSet);
+    public:
+        static std::unique_ptr<FontSet> create(const String& locale, int weight, int width, int slant)
+        {
+            auto* pattern = FcPatternCreate();
```

Comment:
We have `Source/WebCore/platform/graphics/freetype/RefPtrFontconfig.h` which defines the needed method to use `RefPtr<FcPattern>`, so I would prefer to use a smart pointer here:

```suggestion
            RefPtr<FcPattern> pattern = adoptRef(FcPatternCreate());
```

Reply:
And we could add the needed `RefPtr` traits for other Fontconfig types. Some of the API names are misleading and actually add reference counts instead of creating new objects i.e. `FcCharSetCopy()`, and letting our smart pointers take care of the details would avoid that kind of wart.

Note that some Fontconfig types&mdash;typically the ones without a `FcFooCopy` function&mdash;are _not_ reference counted, i.e. `FcFontSet` so for those we would use `std::unique_ptr<>` instead.

Reply:
I thought about it, but since this is the only place where we need to use fontconfig directly I didn't want to depend on anything from platform/freetype (which is indeed confusing, this is not about freetype) in platform/Skia. There are very few manual creare/destroy so I don't think it's worth it. An alternative would be to add our own unique_ptr in this file and use it for all cases, since we don't really need to take references.

Reply:
Right, good point about avoiding depending on `platform/freetype`. I suppose we could move `RefPtrFontconfig.h` into `platform/fontconfig` and use it from different places, but let's land this patch as-is; we can do the cleanup later, _specially_ if we end up needing more code that uses Fontconfig later on.